Key SelectorState on ComponentSelector

jvandort opened the pull request:
We have a hash map of component selectors. The key to the map did not correspond to the the selector it described. The SelectorState is a function of the component selector that it resolves to a component identifier. That should be the only key to the map that caches SelectorState instances.

This commit ensures we avoid re-resolving component selectors by using a cache key that more accurately describes the selector being cached. It also reduces a likely cause of hard-to-find bugs, as the inaccurate cache key that was previously used could cause collisions when dependency substituion was used -- as the prior key used the original requested component selector as key, while the SelectorState resolved the substituted component selector.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

jvandort commented:
Supersedes https://github.com/gradle/gradle/pull/30556

jvandort commented:
@bot-gradle test prf apt

bot-gradle commented:
<!-- {"replyTargetCommentId":2640470212,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_PullRequestFeedback_Trigger","buildId":"94487271","headRef":"a52a60c4445b85ad59bb4f0a34d578027dc3be8a"},{"buildConfigurationId":"Gradle_Master_Check_AllPerformanceTests_Trigger","buildId":"94487286","headRef":"a52a60c4445b85ad59bb4f0a34d578027dc3be8a"}]} -->
I've triggered the following builds:
- [Pull Request Feedback (Trigger) (Check)](https://builds.gradle.org/build/94487271)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=a52a60c4445b85ad59bb4f0a34d578027dc3be8a,PullRequestFeedback&amp;)
- [All Performance Tests (Trigger)](https://builds.gradle.org/build/94487286)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=a52a60c4445b85ad59bb4f0a34d578027dc3be8a,ReadyforRelease&amp;)

bot-gradle commented:
The following builds have passed:
- [All Performance Tests (Trigger)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_AllPerformanceTests_Trigger/94487286)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=a52a60c4445b85ad59bb4f0a34d578027dc3be8a,ReadyforRelease&amp;)

bot-gradle commented:
The following builds have failed:
- [Pull Request Feedback (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_PullRequestFeedback_Trigger/94487271)
  - [Build Scan](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.tags=CI&amp;search.names=gitCommitId&amp;search.values=a52a60c4445b85ad59bb4f0a34d578027dc3be8a&amp;)

jvandort committed:
Remove requested selector from SelectorState

A SelectorState represents a component selector to resolve, after any
substitutions. It has no relationship to the originally requested selector:
a single SelectorState may be shared by many edges, and the requested
selector is a property of each edge, not of the selector being resolved.

Remove SelectorState.getRequested and unify DependencyGraphSelector and
ResolvableSelectorState on a single getComponentSelector accessor. The
remaining consumers of the requested selector were either error messages
or VirtualPlatformState.getForcedVersion, which incorrectly read the
requested version instead of the version the selector actually resolves.

Since resolution failures were previously reported against the requested
selector, EdgeState.failWith now reports the attempted selector and
includes the requested selector as context when the two differ.

jvandort committed:
Decouple SelectorState from the dependencies that use it

A SelectorState may be used by many edges. Previously, it captured the
DependencyState of the first edge that created it, and answered
dependency-level questions from that arbitrary first registrant.

SelectorState now holds only the component selector it resolves.
Dependency-level details -- forcing, locking, changing, the first
dependency artifact, substitution failures -- accumulate as each edge
registers through update(), which is now an unconditional monotonic
merge.

Whether a selector contributes constraint attributes to its target
module is now tracked by counting the constraint edges that currently
use the selector, instead of consulting the first registrant. This
fixes the merged constraint attributes of a module when constraint and
non-constraint edges share a selector: previously the attributes were
either missed entirely or retained after all constraint edges were
gone, depending on which kind of edge registered first. The merged
attributes are now computed lazily and invalidated when a selector
starts or stops being used by constraint edges.

jvandort committed:
Key selector cache on the effective component selector

A SelectorState is a function of the component selector it resolves.
Key the selector cache on that selector alone, deriving the target
module inside the loader so all cached state is a function of the key.
This subsumes the virtual-platform-edge flag, which compensated for
keying on the requested selector, since virtual platform edges cannot
be substituted.

Distinct dependencies can now share a selector, so replace the one-shot
`update` with reference-counted use/release: forcing (hard and soft),
locking, changing, and constraint contributions are counted per edge
and retracted when an edge leaves the graph, triggering reselection on
count transitions.

Also:
- Move substitution failures from the selector to the edge, where they
  are per-dependency, and surface them via EdgeState.getFailure(),
  fixing order-dependent failure masking on shared selectors
- Rename SelectorState.resolved to requiresSelection (inverted), since
  the flag gates module selection, not id resolution
- Selection reasons no longer report "requested" for a forced
  dependency whose version was substituted (test updated)

jvandort commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":5173380431,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_PullRequestFeedback_Trigger","buildId":"115871862","headRef":"2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec"}]} -->
I've triggered the following builds:
- [Pull Request Feedback (Trigger) (Check)](https://builds.gradle.org/build/115871862)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec,PullRequestFeedback&amp;)

bot-gradle commented:
The following builds have passed:
- [Pull Request Feedback (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_PullRequestFeedback_Trigger/115871862)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec,PullRequestFeedback&amp;)

cobexer commented:
@bot-gradle test APT

bot-gradle commented:
<!-- {"replyTargetCommentId":5176803746,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_AllPerformanceTests_Trigger","buildId":"115888273","headRef":"2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec"}]} -->
I've triggered the following builds:
- [All Performance Tests (Trigger)](https://builds.gradle.org/build/115888273)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec,ReadyforRelease&amp;)

bot-gradle commented:
The following builds have passed:
- [All Performance Tests (Trigger)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_AllPerformanceTests_Trigger/115888273)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=2e90c760dc79b8f842c1191d2a9ab0d9ebff80ec,ReadyforRelease&amp;)
